### PR TITLE
Typescript/tab controller move to typescript

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
TabController - move to typescript
This is step one from 3 (step 2 is move to functional component and 3 is to add a presenter for the centering).
There are a few TODOs for typescript, some will be removed once the move to functional component is done, but I'm not sure all will be removed, especially the ones that are related to reanimated, they don't export all of their types.

## Changelog
TabController - move to typescript

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -36,6 +36,7 @@ export {default as Overlay, OverlayTypes} from './components/overlay';
 export {default as RadioButton, RadioButtonPropTypes, RadioButtonProps} from './components/radioButton/RadioButton';
 export {default as RadioGroup, RadioGroupPropTypes, RadioGroupProps} from './components/radioButton/RadioGroup';
 export {default as Switch, SwitchProps} from './components/switch';
+export {default as TabController, TabControllerProps, TabBarItemProps} from './components/tabController';
```
Comment (ethanshar) [inline src/index.ts:39]: `TabBarItemProps` might conflict in the future with our TabBar component which will probably have also TabBatItemProps :/ 
Maybe we can name this one TabControllerItemProps, WDYT?

File: src/components/tabController/PageCarousel.tsx [hunk 2]
```diff
@@ -23,16 +23,16 @@ class PageCarousel extends PureComponent {
...
   }
 
-  onTabChange = ([index]) => {
-    this.scrollToPage(index, true);
+  onTabChange = ([index]: readonly number[]) => {
+    this.scrollToPage(index);
   };
 
-  scrollToPage = (pageIndex, animated) => {
+  scrollToPage = (pageIndex: number) => {
```
Comment (ethanshar) [inline src/components/tabController/PageCarousel.tsx:35]: It seems like you changes the functionality here. 
Why did you remove the `animated` parameter?

File: src/components/tabController/TabBar.tsx [hunk 3]
```diff
@@ -241,20 +284,21 @@ class TabBar extends PureComponent {
...
         if (Constants.isRTL && Constants.isAndroid) {
           const scrollingWidth = Math.max(0, this.contentWidth - this.containerWidth);
           targetOffset = scrollingWidth - targetOffset;
         }
 
-        this.tabBar.current.scrollTo({x: targetOffset, animated});
+        if (this.tabBar && this.tabBar.current) {
+          this.tabBar.current.scrollTo({x: targetOffset, animated});
+        }
       }
```
Comment (ethanshar) [inline src/components/tabController/TabBar.tsx:297]: Consider using optional chaining here (:

File: src/components/tabController/TabBarItem.tsx [hunk 4]
```diff
@@ -0,0 +1,325 @@
+// TODO: support commented props
```
Comment (ethanshar) [inline src/components/tabController/TabBarItem.tsx:1]: Do you think it's possible to fix this file push so it won't be consider as a new file and will replace the original JS file.
If it's too complicated then never mind.. But it'll be nice not to lost git history for this file.

File: src/components/tabController/TabPage.tsx [hunk 5]
```diff
@@ -68,7 +72,7 @@ export default class TabPage extends PureComponent {
     const {targetPage} = this.context;
     const {index, lazy} = this.props;
     return block([
-      cond(and(eq(targetPage, index), lazy, eq(this._loaded, 0)), [set(this._loaded, 1), call([], this.lazyLoad)]),
+      cond(and(eq(targetPage, index), Number(lazy), eq(this._loaded, 0)), [set(this._loaded, 1), call([], this.lazyLoad)]),
```
Comment (ethanshar) [inline src/components/tabController/TabPage.tsx:75]: Did you find a bug here? Does the number conversion here actually fix something?

File: src/components/tabController/index.tsx [hunk 6]
```diff
@@ -204,4 +237,13 @@ TabController.TabBar = TabBar;
...
 TabController.PageCarousel = PageCarousel;
-export default asBaseComponent(TabController);
+export default asBaseComponent<
+  TabControllerProps,
+  {
+    TabBar: typeof TabBar;
+    TabBarItem: typeof TabBarItem;
+    TabPage: typeof TabPage;
+    PageCarousel: typeof PageCarousel;
+  }
```
Comment (ethanshar) [inline src/components/tabController/index.tsx:247]: What about doing `typeof TabController` like we usually do. 
Wouldn't that work?

**Review by ethanshar — COMMENTED** [review]
Overall look good. 
Wrote some comments. 

Anyway, i Noticed some performance issues with the component, but I think it has nothing to do with your PR. 
Recently, we update reanimated version to v13.1 and I noticed some weird issues in TabController. 
Do you experience them as well?

File: src/index.ts (hunk 1 above)
Comment (M-i-k-e-l) [reply src/index.ts:39]: I think it might get confusing since all the sub components there are (for TabController), and I'm not sure people won't need their props as well.
Maybe we can add them statically? Either on the component or create another "MegaProp" that has them all?
So either TabController.Props, TabController.TabBar.Props... or TabControllerProps.Main, TabControllerProps.TabBar...
Not sure how easy it is to do or understand, WDYT?

File: src/components/tabController/PageCarousel.tsx (hunk 2 above)
Comment (M-i-k-e-l) [reply src/components/tabController/PageCarousel.tsx:35]: It was not used, `false` was always sent (and I had an error about it).

**Comment by M-i-k-e-l** [comment]
> Overall look good.
> Wrote some comments.
> 
> Anyway, i Noticed some performance issues with the component, but I think it has nothing to do with your PR.
> Recently, we update reanimated version to v13.1 and I noticed some weird issues in TabController.
> Do you experience them as well?

Yes, I think I know what you mean, it also sometimes does not go to the correct page. I was thinking that it might be solved with the new logic 🤞

File: src/components/tabController/TabBar.tsx (hunk 3 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabBar.tsx:297]: Fixed :)

File: src/components/tabController/TabBarItem.tsx (hunk 4 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabBarItem.tsx:1]: Looking in VSCode I have the history, where are you looking?

File: src/components/tabController/TabPage.tsx (hunk 5 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabPage.tsx:75]: I think it was for typescript, but you can see it's the same on line 53.

File: src/components/tabController/index.tsx (hunk 6 above)
Comment (M-i-k-e-l) [reply src/components/tabController/index.tsx]: Yes, my mistake, fixed

**Comment by ethanshar** [comment]
> Yes, I think I know what you mean, it also sometimes does not go to the correct page. I was thinking that it might be solved with the new logic 🤞
> Edit: looking at it again, not sure the new logic will fix it since it happens not only in centered mode.

Ok.. so I just verified.
On react-native-reanimated@1.9.0 (the previous version we had), everything works great
After updating to react-native-reanimated@1.13.1 (which was required for something else) things started to go south.. 

So I suggest that after we merge this PR, We'll investigate the issue first before going forward.
Usually when there are issues because of version upgrade, it's probably something small we need to tweak. 🤞

**Comment by M-i-k-e-l** [comment]
> > Yes, I think I know what you mean, it also sometimes does not go to the correct page. I was thinking that it might be solved with the new logic 🤞
> > Edit: looking at it again, not sure the new logic will fix it since it happens not only in centered mode.
> 
> Ok.. so I just verified.
> On react-native-reanimated@1.9.0 (the previous version we had), everything works great
> After updating to react-native-reanimated@1.13.1 (which was required for something else) things started to go south..
> 
> So I suggest that after we merge this PR, We'll investigate the issue first before going forward.
> Usually when there are issues because of version upgrade, it's probably something small we need to tweak. 🤞

ok

File: src/components/tabController/TabPage.tsx (hunk 5 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx:75]: 👍

File: src/components/tabController/TabBarItem.tsx (hunk 4 above)
Comment (ethanshar) [reply src/components/tabController/TabBarItem.tsx:1]: In github? 
This file is considered new, so it's a little difficult to spot what has changed.. 
It's not big deal, but if you can easily fix it, that'll be great..

File: src/components/tabController/PageCarousel.tsx (hunk 2 above)
Comment (ethanshar) [reply src/components/tabController/PageCarousel.tsx:35]: ok.

File: src/index.ts (hunk 1 above)
Comment (ethanshar) [reply src/index.ts:39]: I'm not sure you can create this type of nesting in typings... 
If it's possible then let's do it.. 

I guess that's the reason other libraries' typings have weird long names..

File: src/components/tabController/TabBarItem.tsx (hunk 4 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabBarItem.tsx:1]: Weird, I did the same thing I did with other files (`git mv <old_file> <new_file>`)...
And I do see the history in VSCode, I think we'll 🤞 for it to be ok?

File: src/components/tabController/index.tsx [hunk 7]
```diff
@@ -204,4 +237,8 @@ TabController.TabBar = TabBar;
 TabController.TabBarItem = TabBarItem;
 TabController.TabPage = TabPage;
 TabController.PageCarousel = PageCarousel;
-export default asBaseComponent(TabController);
+export default asBaseComponent<
+  TabControllerProps,
+  typeof TabController
+  //@ts-ignore typescript - will be fixed when moved to functional component
+>(TabController);
```
Comment (ethanshar) [inline src/components/tabController/index.tsx:244]: no need for asBaseComponent

Comment (M-i-k-e-l) [reply src/components/tabController/index.tsx:244]: It seems it was `asBaseComponent` to begin with, I see we're using its `config` in private for `carouselPageWidth`.

Comment (ethanshar) [reply src/components/tabController/index.tsx:244]: you are completely right, I accidentally looked at the old TabController file that sits under Incubator 🤦‍♂️

File: src/index.ts (hunk 1 above)
Comment (M-i-k-e-l) [reply src/index.ts]: I went with `TabControllerProps` namespace, let me know if that's ok

File: src/components/tabController/index.tsx (hunk 7 above)
Comment (M-i-k-e-l) [reply src/components/tabController/index.tsx]: Yes, tell me about it, I spent some time trying to figure about why my changes are not applied in VSCode (why are my files still in javascript)... :)

File: src/index.ts (hunk 1 above)
Comment (ethanshar) [reply src/index.ts]: The namespace is nice, but now users will need to do  this?

`TabControllerProps.TabControllerProps`? it's a little weird don't you think?

File: src/components/tabController/index.tsx (hunk 7 above)
Comment (ethanshar) [reply src/components/tabController/index.tsx]: yes, we plan to remove these files in V6. 
My bad, I should have warn you about it..

Comment (M-i-k-e-l) [reply src/components/tabController/index.tsx]: No worries :)

File: src/index.ts (hunk 1 above)
Comment (M-i-k-e-l) [reply src/index.ts]: :s

Any suggestions? `TabControllerProps.Props` or maybe create the `namespace` as `TabController` and everything will be inside it (but the it'll be `TabController.TabController` and worse `TabController.TabController.TabBar`).

Comment (ethanshar) [reply src/index.ts]: TBH, Im starting to think we should drop the namespace thing, it's a little messy and too nested. 
Im not sure people are used to import typings this way.
I prefer to have 
TabControllerProps
TabControllerBarProps
TabControllerItemProps
TabControllerPageProps, etc...

It's not the best, but feels simpler.. WDYT?

Comment (M-i-k-e-l) [reply src/index.ts]: I agree, changed as you recommended

**Review by ethanshar — APPROVED** [review]
